### ISSUE TITLE
CollectionExtensions is missing optimizations for Dictionary

### ISSUE THREAD

**Issue by marek-safar** [dotnet/corefx#20555]
New CollectionExtensions methods Remove and TryAdd could use Dictionary<,> implementation directly when input value is of Dictionary<,> type.

**Comment by kellypleahy** [comment]
Is this really an optimization?  Is the type comparison and cast going to be just as slow as the two public calls, or do you mean we should have extension methods that take the concrete `Dictionary<TKey,TValue>` type as an argument?

**Comment by marek-safar** [comment]
The optimization is to replace 2 calls with type check and 1 call. For example Dictionary<,> implements TryAdd which is much faster than calling Contains+Add in CollectionExtensions implementation

**Comment by stephentoub** [comment]
> The optimization is to replace 2 calls with type check and 1 call.

If it's a Dictionary<,> and if it doesn't contain the item being added.  If it's not a Dictionary<,>, it's adding a type check and nothing is saved.  And if it is a Dictionary<,> but the item already exists, it's replacing 1 interface call with 1 type check and 1 call.  Those could very well be the right trade-offs, since Dictionary<,> is so common, but it would need to be investigated and understood before making such a change.  If someone really cares about this kind of throughput accessing a dictionary, they'd likely want to avoid going through an interface in general.

**Comment by kellypleahy** [comment]
@stephentoub yes, well put.  That was exactly what I was trying to ask, as it didn't seem obvious to me that the tradeoff of doing the type check 100% of the time was worthwhile, when the user could get the perf gain that wanted just by using the concrete type.
